src/visualisation/geoviewer.py

```diff
@@ -0,0 +1,414 @@
+            if name is None:
+                name = layer["name"]
+            layer = ipyleaflet.basemap_to_tiles(layer)
+        else:
+            if name is None:
+                name = layer.name
+        if layer.model_id in self._layer_ids or name in self.layer_dict["maps"].keys():
+            raise ipyleaflet.LayerException(
+                "layer with same name already on map, change name argument: %r" % layer
+            )
+
+        self.layer_dict["maps"][name] = dict(map=dict(layer=layer, active=True))
+        self.layer_update()
+
+    def add_graph(self, graph: geograph.GeoGraph, name: str = "Graph") -> None:
+        """Add GeoGraph to viewer.
+
+        Args:
+            graph (geograph.GeoGraph): graph to be added
+            name (str, optional): name shown in control panel. Defaults to "Graph".
+        """
+        if name in graph.habitats.keys():
+            raise ValueError(
+                "Name given cannot be same as habitat name in given GeoGraph."
+            )
+        if name in self.layer_dict["graphs"]:
+            raise ValueError(
+                "Graph with the same name already added to GeoGraphViewer."
+            )
+
+        graphs = {name: graph, **graph.habitats}
+
+        for current_name, current_graph in graphs.items():
+
+            nx_graph = current_graph.graph
+            is_habitat = not current_name == name
+
+            # Creating layer with geometries representing graph on map
+            nodes, edges = graph_utils.create_node_edge_geometries(
+                nx_graph, crs=current_graph.crs
+            )
+            graph_geometries = pd.concat([edges, nodes]).reset_index()
+            graph_geo_data = ipyleaflet.GeoData(
+                geo_dataframe=graph_geometries.to_crs(self.gpd_crs_code),
+                name=current_name + "_graph",
+                **self.layer_style["graph"]
+            )
+
+            # Creating choropleth layer for patch polygons
+            pgon_df = current_graph.df.loc[list(nx_graph)]
+            pgon_choropleth = self._get_choropleth_from_df(pgon_df)
+
+            # Creating layer for graph components
+            component_df = current_graph.get_graph_components(
+                calc_polygons=True
+            ).df.copy()
+            if is_habitat:
+                component_df.geometry = component_df.geometry.buffer(
+                    current_graph.max_travel_distance
+                )
+            component_choropleth = ipyleaflet.GeoData(
+                geo_dataframe=component_df.to_crs(WGS84),
+                name=current_name + "_components",
+                **self.layer_style["components"]
+            )
+
+            # Creating layer for disconnected (no-edge) nodes
+            disconnected_nx_graph = nx_graph.subgraph(
+                [node for node in nx_graph.nodes() if nx_graph.degree[node] == 0]
+            )
+
+            discon_nodes, _ = graph_utils.create_node_edge_geometries(
```

Comment:
Can we not use `include_edges=False` here to save computing the edges?

Reply:
And the same applies to the call below as well.

Reply:
Yeah, good point. I implemented that argument for those two calls but then didn't use it.